sklearn/ensemble/_gradient_boosting.pyx

```diff
@@ -84,22 +88,96 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
+                                           np.ndarray[float64, ndim=2] out):
+    """Predicts output for regression tree inplace and adds scaled value to ``out[i, k]``.
+
+    The function assumes that the ndarray that wraps ``X`` is csr_matrix.
+    """
+    cdef DTYPE_t* X_data = <DTYPE_t*>(<np.ndarray> X.data).data
+    cdef INT32_t* X_indices = <INT32_t*>(<np.ndarray> X.indices).data
+    cdef INT32_t* X_indptr = <INT32_t*>(<np.ndarray> X.indptr).data
+
+    cdef SIZE_t n_samples = X.shape[0]
+    cdef SIZE_t n_features = X.shape[1]
+    cdef SIZE_t n_stages = estimators.shape[0]
+    cdef SIZE_t n_outputs = estimators.shape[1]
+
+    # Initialize output
+    cdef float64* out_ptr = <float64*> out.data
+
+    # Indices and temporary variables
+    cdef SIZE_t sample_i = 0
+    cdef SIZE_t feature_i = 0
+    cdef SIZE_t stage_i = 0
+    cdef SIZE_t output_i = 0
+    cdef Node *root_node = NULL
+    cdef Node *node = NULL
+    cdef double *value = NULL
+
+    # Initialize auxiliary data-structure
+    cdef DTYPE_t feature_value = 0.
+    cdef DTYPE_t* X_sample = NULL
+
+    # feature_to_sample as a data structure records the last seen sample
+    # for each feature; functionally, it is an efficient way to identify
+    # which features are nonzero in the present sample.
+    cdef SIZE_t* feature_to_sample = NULL
+
+    safe_realloc(&X_sample, n_features * sizeof(DTYPE_t))
+    safe_realloc(&feature_to_sample, n_features * sizeof(SIZE_t))
+
+    memset(feature_to_sample, -1, n_features * sizeof(SIZE_t))
+
+    # Cycle through all samples
+    for sample_i in range(n_samples):
+        for feature_i in range(X_indptr[sample_i], X_indptr[sample_i + 1]):
+            feature_to_sample[X_indices[feature_i]] = sample_i
+            X_sample[X_indices[feature_i]] = X_data[feature_i]
+
+        # Cycle through all stages
+        for stage_i in range(n_stages):
+            # Cycle through all trees
+            for output_i in range(n_outputs):
+                root_node = (<Tree> estimators[stage_i, output_i].tree_).nodes
+                value = (<Tree> estimators[stage_i, output_i].tree_).value
+                node = root_node
+
+                # While node not a leaf
+                while node.left_child != TREE_LEAF:
+                    # ... and node.right_child != TREE_LEAF:
+                    if feature_to_sample[node.feature] == sample_i:
+                        feature_value = X_sample[node.feature]
+                    else:
+                        feature_value = 0.
+
+                    if feature_value <= node.threshold:
+                        node = root_node + node.left_child
+                    else:
+                        node = root_node + node.right_child
+                out_ptr[sample_i * n_outputs + output_i] += (scale
+                    * value[node - root_node])
+
+    # Free auxiliary arrays
+    free(X_sample)
+    free(feature_to_sample)
+
+
```

Comment:
@raghavrv, ok, I'll try to test it with turned off checks. But seems that most time-consuming checks are already disabled:

    # cython: cdivision=True
    # cython: boundscheck=False
    # cython: wraparound=False

It's actually interesting why sparse is 5 times slower, because I did my own benchmarks (Code is provided above), it was slower, but not even 2 times slower.